### PR TITLE
[MRG] Implement Score so we can perform Gridsearch

### PR DESCRIPTION
**Pull request by pmyers16**
Thanks for contributing. If this is your first time,
make sure to read [contributing.md](https://github.com/adma2392/mne-hfo/blob/master/CONTRIBUTING.md)

PR Description
--------------

Closes: #14 

Implements various scoring mechanisms so that we can perform SKLearn's GridSearchCV.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/adam2392/mne-hfo/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: mne_hfo/posthoc.py [hunk 1]
```diff
@@ -338,89 +338,69 @@ def _find_overlapping_events(list1, list2):
...
-    ytrue_df[bn] = ytrue_df[bn].apply(pd.to_numeric())
-    ypredict_df[bn] = ypredict_df[bn].apply(pd.to_numeric())
-
-    # If df has duration instead of end time, add a new column
-    if bn[1].lower() == "duration":
-        ytrue_df = _append_offset_to_df(ytrue_df, bn)
-        ypredict_df = _append_offset_to_df(ypredict_df, bn)
-        bn[1] = "offset"
-    match_df = pd.DataFrame(columns=('gs_index', 'dd_index'))
+    dc = ["onset", "duration"]
```
Comment (adam2392) [inline mne_hfo/posthoc.py:347]: Eventually will need a description of match-true, match-pred and match-total.

File: mne_hfo/posthoc.py [hunk 2]
```diff
@@ -338,89 +338,69 @@ def _find_overlapping_events(list1, list2):
...
-    freq_name: str
-        Name of frequency column
-    sec_unit: int
-        Number representing one second of signal - this can
-        significantly improve the speed of this function
-    sec_margin: int
-        Margin for creating subsets of compared data - should be set according
-        to the length of compared events (1s for HFO should be enough)
+def match_detections(ytrue_df, ypredict_df, label:str=None, sec_margin: float=1., method="match-true"):
+    # Calculate frq from dataframe, then matching window
```
Comment (adam2392) [inline mne_hfo/posthoc.py:342]: Eventually will need a description of match-true, match-pred and match-total.

Comment (adam2392) [reply mne_hfo/posthoc.py:342]: Also describe what is returned. I'm assuming a structured dataframe, or is it a list?

**Review by adam2392 — COMMENTED** [review]
The question I have is how do we go from:

```
expected_dict_true = {
        "true_index": [0, 1, 2, 3],
        "pred_index": [0, 1, None, 2]
    }
```
to a prediction score? Does one just convert true_index -> ytrue (all 1's). How can we encompass a false-negative (i.e. an HFO that is not actually present in the reference dataset)?

**Comment by adam2392** [comment]
true positive rate (recall), false discovery rate, precision, kappa score

File: mne_hfo/scores.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    float
+
+    """
+    from mne_hfo import match_detections
+    overlap_df = match_detections(y, y_pred, method="match-total")
+    tp, fp, fn = _calculate_match_stats(overlap_df)
+    return fp / (fp + tp)
+
+
+def accuracy(y, y_pred):
```
Comment (adam2392) [inline mne_hfo/scores.py:91]: All of these functions require certain columns inside, so this should be error checked.

File: mne_hfo/scores.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    Calculate accuracy as: accuracy = tp / (tp + fp + fn).
+
+    Follows usual formula for accuracy: accuracy = (tp + tn) / (tp + tn + fp + fn)
+    but assumes tn = 0.
+
+    Parameters
+    ----------
+    y : pd.DataFrame
+        Annotation Dataframe with actual labels
```
Comment (adam2392) [inline mne_hfo/scores.py:101]: Technically, DataFrame with `ypred index` and ytrue index columns that mean ... 
not the Annotation Dataframe.

**Review by adam2392 — CHANGES_REQUESTED** [review]
In ` tests/test_posthoc.py`, I think now we can add a test:

```Python
from sklearn.metrics import make_scorer

@pytest.parametrize(scorer, [<the 3 different scores>])
def test_hyperparameter_search_cv(scorer):
parameters = {'threshold': [1,2,3], ...}
detector = SomeDetector()
scorer = make_scorer(scorer, <arguments>)
cv = GridSearchCV(detector, parameters, scoring=scorer, cv=dummycv)

# use test dataset and trim down to like... 10 seconds and maybe add 1-2 artifical HFOs
cv.fit(raw, true_hfo_df)

# add checks
```

To make a dummy cv object: https://stackoverflow.com/questions/44636370/scikit-learn-gridsearchcv-without-cross-validation-unsupervised-learning

File: mne_hfo/scores.py (hunk 4 above)
Comment (pmyers16) [reply mne_hfo/scores.py:103]: These are the annotation dataframes. Matching occurs in the scoring function. Would you prefer another way?

Comment (pmyers16) [reply mne_hfo/scores.py:103]: Sorry these are event dataframes, I will change

Comment (adam2392) [reply mne_hfo/scores.py:103]: Sorry I meant that these can be annotation data frames (not events please for the sake of consistency) but that they don't really require anything beyond two columns right?

These two columns encode some sort of data that allows you to then compute said metric. 

So it would be good to have a description of what those two columns are. Can be copy pasted

**Comment by adam2392** [comment]
# Making SearchCV work
So in order to make GridSearch work, we need to change the way we structure input into the `fit()` function. Some minor changes:

1. For raw data: we want channels to correspond to `n_samples` in sklearn, and time points to correspond to `n_features`.
2. For "y" labeled data: we want a list that is `n_samples` long. That is each channel will get a list now of HFO endpoints. Example:

```Python
# has length equal to number of channels
# the second channel has no HFOs detected
y = [[(0, 20), (10, 40), ...], [], [(20, 30), (10, 20), ...], ...]]  
```

For example, try out the `test_hyperparameter_search_cv` with accuracy just returning dummy "1" rn. It should work. 

# Suggestions and comments
In order to make code easier to maintain and build upon and testable, here are a few comments:

1. Only use `annotations.tsv style DataFrame`: we aren't using the events dataframe format. I simply have that for legacy pulling over from the Fedele / other datasets that store things in the events.tsv files. Might even make those private eventually. In all dataframe functions, assume annotations because one then has the channels column. 
2. `match_detections` does not fit into the above structure: I know it's a "functional" working fucntion, but I can't figure out how it works. I think this needs to be refactored to fit in line with the working data structure assumptions of the rest of the repository (i.e. annotations-style DataFrame). 
3. As a result, of the above 2, the scoring functions are not easy to work with if one starts from Annotations DataFrames. Ideally, one can just feed in 2 annotations.tsv files formatted according to BIDS and a score can be computed. 

I made a new `match_detected_annotations` function inside `posthoc.py:L341`, which I would like you to implement and test. Once you have that working, then everything else should work.

# Next Steps

- [ ] Implement `match_detected_annotations`
- [ ] Test `match_detected_annotations`
- [ ] Implement scoring functions (FYI, see `_compute_score_data()` function.

File: mne_hfo/scores.py [hunk 5]
```diff
@@ -119,50 +115,47 @@ def accuracy(y, y_pred):
...
-    y_pred_bool = y_pred_series.notna().to_list()
-    return y_true_bool, y_pred_bool
+def _compute_score_data(y, y_pred, method):
+    """Compute basic HFO scoring metrics."""
+    from mne_hfo import match_detections
+    # if isinstance(y, pd.DataFrame):
+    #     y = _check_df(y, df_type='annotations')
+    # else:
+    # assume y is now in the form of list of (onset, offset) per channel
+    # y = _make_ydf_sklearn(y, ch_names=)
```
Comment (pmyers16) [inline mne_hfo/scores.py:133]: Im confused by your psuedocode here. If y is a df, the check that it is valid. If it is not a df, then pass it to _make_ydf_sklearn. But that function takes a df (which we just checked that it isnt) and outputs a list of list of tuples. Wouldn't we want it the other way around? If a list of list of tuples is passed in, convert to a dataframe that we can match?

Comment (adam2392) [reply mne_hfo/scores.py:133]: Oh yeah sorry this isn't pseudocode, just didn't have anything here.

Once `match_annotations` is done, then this should as you say:
- check if it's a dataframe, if so, check that it's the correct type of dataframe
- if it's a list, assume that it's a list of HFO events, which needs to then get converted to an annotations dataframe. In doing so, one needs to pass in the `ch_names` that corresponds to the list of HFO events. 

^ the fact that you need to pass in a list, is a bit lame, so I'm wondering if instead we should be be making `_make_y_sklearn` output an OrderedDictionary.

Comment (pmyers16) [reply mne_hfo/scores.py:133]: So we can handle this two ways. Have the scoring functions take kwargs of ch_name and sfreq, or use an OrderedDictionary where the keys are ch_names and sfreq is some dummy value like 1 just to get around the requirement since its not actually used. Or I suppose we could use an OrderedDictionary with sfreq as a kwarg

**Comment by adam2392** [comment]
I made some fixes (minor), such as removing the `dc=['onset', 'offset']` stuff. Just explicitly set these columns as numeric, and compute `offset = onset + duration`. Each dataframe can be assumed to be in `Annotations` format.

- I'm going to remove the need for `ch_names` in the scoring, and add a private function to go from OrderedDictionary of sklearn compatible data structure -> Annotations DF.

### A few questions:

1. I'm not sure why you had to add the `filter_band` to LineLengthDetector in the test
2. Why did you have to add `sfreq` to the scorers?

**Comment by adam2392** [comment]
Can you update `whats_new.rst` and then this should be good to go?

Else if you also want to add the additional step to the notebook showing the gridsearchCV performance on the reference Fedele dataset, then you can do that here.

File: tests/test_posthoc.py [hunk 6]
```diff
@@ -103,3 +283,69 @@ def test_metrics_df(end_sec, rate):
     # error if specifying channel names are not inside dataframe
     with pytest.raises(ValueError, match=''):
         compute_chs_hfo_rates(annot_df, rate=rate, ch_names=['A0', 'A1'])
+
+# @pytest.mark.skip()
```
Comment (adam2392) [inline tests/test_posthoc.py:287]: Let's just delete all this now, that we have a tested and working version of match_annotations

**Review by adam2392 — CHANGES_REQUESTED** [review]
Can you just migrate all cells for running on simulation to another notebook under `tutorials/` and rename "basic_compute_hfos.ipynb" to "compute_hfos_on_fedele_dataset.ipynb"?

**Review by adam2392 — APPROVED** [review]
